src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -70,22 +72,31 @@ public JsonClassInfo ElementClassInfo
 
         public void UpdateSortedPropertyCache(ref ReadStackFrame frame)
         {
-            // Check if we are trying to build the sorted cache.
-            if (frame.PropertyRefCache == null)
-            {
-                return;
-            }
+            Debug.Assert(frame.PropertyRefCache != null);
 
-            List<PropertyRef> newList;
-            if (_propertyRefsSorted != null)
-            {
-                newList = new List<PropertyRef>(_propertyRefsSorted);
-                newList.AddRange(frame.PropertyRefCache);
-                _propertyRefsSorted = newList.ToArray();
-            }
-            else
+            lock (_cacheLock)
             {
-                _propertyRefsSorted = frame.PropertyRefCache.ToArray();
+                if (_propertyRefsSorted != null)
+                {
+                    List<PropertyRef> replacementList = new List<PropertyRef>(_propertyRefsSorted);
+                    Debug.Assert(replacementList.Count <= PropertyNameCountCacheThreshold);
+
+                    // Verify replacementList will not become too large. This could occur in race conditions.
+                    List<PropertyRef> listToAppend = frame.PropertyRefCache;
+                    while (replacementList.Count + listToAppend.Count > PropertyNameCountCacheThreshold)
+                    {
+                        listToAppend.RemoveAt(listToAppend.Count - 1);
+                    }
+
+                    // Add the new items; we don't check for duplicates so in a race condition duplicates are possible
+                    // however that is tolerated during property lookup.
+                    replacementList.AddRange(listToAppend);
+                    _propertyRefsSorted = replacementList.ToArray();
```

Comment:
_propertyRefsSorted should be marked as `volatile`.

Reply:
It was previously, but changed it based on assumption: if a given thread 1 accesses _propertyRefsSorted and it changes by thread 2, is the memory accessed by thread 1 still valid? (i.e. not GC'd). If so, then `volatile` should not be necessary since the variable won't be "cached" long enough to affect perf.

Reply:
With or without volatile won't affect the type / memory safety of the program.  What it will impact is instruction ordering and whether/how writes from one thread are observed by another.  Writes from one thread can end up being seen in different orders by other threads.  The classic case of this (as it could apply here) is one thread creating an array, populating it by writing into the array, and then publishing that array by writing the reference to it to a shared location.  But another thread need not witness those writes in the same order... it could see the write of the array to the field before it sees the writes of the data into the array, which means it could end up reading from an unpopulated or only partially populated data structure.  By putting volatile on the field, you give reads acquire semantics and writes release semantics, which avoids such ordering issues in these common patterns.

Reply:
OK thanks. Given that this cache is always "consistent" and has a fallback of the "real" metadata, I do not believe volatile is needed. However based on other comments about using a local refence to the cache during reading, the volatile is probably necessary to ensure the local is always used.